controller/rest/longpoll.go

```diff
@@ -52,54 +58,77 @@ func (j *longpollOnceJob) Poll() (interface{}, time.Duration) {
 	}
 }
 
+type PollWorkItem struct {
+	key  interface{}
+	args interface{}
+	task longpollOnceTask
+}
+
 type longpollOnceMgr struct {
-	mux     sync.Mutex
-	max     int
-	timeout time.Duration
-	linger  time.Duration
-	jobs    map[interface{}]*longpollOnceJob
+	mux       sync.RWMutex
+	poolSize  int
+	timeout   time.Duration
+	linger    time.Duration
+	jobs      map[interface{}]*longpollOnceJob // use as a queue
+	workQueue chan PollWorkItem
 }
 
-func NewLongPollOnceMgr(timeout, linger time.Duration, max int) *longpollOnceMgr {
-	return &longpollOnceMgr{
-		max:     max,
-		timeout: timeout,
-		linger:  linger,
-		jobs:    make(map[interface{}]*longpollOnceJob),
+func (m *longpollOnceMgr) pollWorker() {
+	for workItem := range m.workQueue {
+		m.mux.RLock()
+		job, exists := m.jobs[workItem.key]
+		m.mux.RUnlock()
+
+		if exists {
+			job.dataCh <- workItem.task.Run(workItem.args)
+			// In case polling returned, wait a while before discarding the record
+			select {
+			case <-time.After(m.linger):
+			case <-job.polled:
+			}
```

Comment:
Why is there a need to wait before removing a completed job?

Reply:
I think you are right, the existing design to make sure the job delete after some second => which should wait for the job to be done.

we should remove this.